docs/csharp/nullable-warnings.md

```diff
@@ -0,0 +1,130 @@
+- Apply the null forgiving operator `!` to the expression to force the state to *not-null*.
+
+## Possible null assigned to a nonnullable reference
+
+The compiler emits these warnings when you attempt to assign an expression that is *maybe-null* to a variable that is nonnullable. For example:
+
+:::code language="csharp" source="./snippets/null-warnings/Program.cs" id="PossibleNullAssignment":::
+
+You can take one of three actions to address these warnings. One is to add the `?` annotation to make the variable a nullable reference type. That change may cause other warnings. Changing a variable from a non-nullable reference to a nullable reference changes its default *null-state* from *not-null* to *maybe-null*. The compiler's static analysis may find instances where you dereference a variable that is *maybe-null*.
+
+The other actions instruct the compiler that the right-hand-side of the assignment is *not-null*. The expression on the right-hand-side could be null checked before assignment, as shown in the following example:
+
+:::code language="csharp" source="./snippets/null-warnings/Program.cs" id="NullGuard":::
+
+The previous examples demonstrate assignment to the return value of a method. You may annotate the method (or property) to indicate when a method returns a not-null value. The <xref:System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute?displayProperty=nameWithType> often specifies that a return value is *not-null* when an input argument is *not-null*. Another alternative is to add the null forgiving operator, `!` to the right-hand side:
+
+```csharp
+string msg = TryGetMessage(42)!;
+```
+
+Fixing a warning for assigning a *maybe-null* expression to a *not-null* variable involves one of four techniques:
+
+- Change the left side of the assignment to a nullable type. This action may introduce new warnings when you dereference that variable.
+- Provide a null-check before the assignment.
+- Annotate the API that produces the right-hand side of the assignment.
+- Add the null forgiving operator to the right-hand side of the assignment.
+
+## Nonnullable reference not initialized
+
+Other warnings are generated when a nonnullable reference variable isn't initialized when declared, or in a constructor. Consider the following class as an example:
+
+:::code language="csharp" source="./snippets/null-warnings/PersonExamples.cs" id="PersonExample":::
+
+Neither `FirstName` nor `LastName` are guaranteed initialized. If this code is new, consider changing the public interface. The above example could be updated as follows:
+
+:::code language="csharp" source="./snippets/null-warnings/PersonExamples.cs" id="WithConstructor":::
+
+If you require creating a `Person` object before setting the name, you can initialize the properties using a default non-null value:
+
+:::code language="csharp" source="./snippets/null-warnings/PersonExamples.cs" id="Initializer":::
+
```

Comment:
I'm curious, is there ever a valid use case for using the null forgiving operator to say that you're ok with it being `null` to start out, even though you define the type as non-nullable?

```csharp
public class Person
{
    public string FirstName { get; set; } = null!;
    public string LastName { get; set; } = null!;
}
```

Reply:
I'd rather not recommend it. It's valid, but you're telling the compiler "I trust all callers to do the right thing." I might use it it legacy code, where I don't want to introduce breaking changes. Otherwise, no.